openjdk/mmtkUpcalls.cpp

```diff
@@ -111,30 +111,37 @@ static void mmtk_spawn_gc_thread(void* tls, int kind, void* ctx) {
 
 static void mmtk_block_for_gc() {
   MMTkHeap::heap()->_last_gc_time = os::javaTimeNanos() / NANOSECS_PER_MILLISEC;
-  log_debug(gc)("Thread (id=%d) will block waiting for GC to finish.", Thread::current()->osthread()->thread_id());
 
   // We must read the counter before entering safepoint.
-  // This thread has just triggered GC.
-  // Before this thread enters safe point, the GC cannot start, and therefore cannot finish,
-  // and cannot increment the counter mmtk_start_the_world_count.
-  // Otherwise, if we attempt to acquire the gc_lock first, GC may have triggered stop-the-world
-  // first, and this thread will be blocked for the entire stop-the-world duration before it can
-  // get the lock.  Once that happens, the current thread will read the counter after the GC, and
-  // wait for the next non-existing GC forever.
+  // This thread (or another mutator) has just triggered GC.
+  // The GC cannot start until all mutators enter safepoint.
+  // If the GC cannot start, it cannot finish, and cannot increment mmtk_start_the_world_count.
+  // Otherwise, if we enter safepoint before reading mmtk_start_the_world_count,
+  // we will allow the GC to start before we read the counter,
+  // and the GC workers may run so fast that they have finished one whole GC and incremented the
+  // counter before this mutator reads the counter for the first time.
+  // Once that happens, the current mutator will wait for the next GC forever,
+  // which may not happen at all before the program exits.
   size_t my_count = Atomic::load(&mmtk_start_the_world_count);
   size_t next_count = my_count + 1;
 
+  log_debug(gc)("Will block until the start_the_world counter reaches %zu.", next_count);
+
   {
-    // Once this thread acquires the lock, the VM will consider this thread to be "in safe point".
-    MutexLocker locker(MMTkHeap::heap()->gc_lock());
+    // Enter safepoint.
+    JavaThread* thread = JavaThread::current();
+    ThreadBlockInVM tbivm(thread);
```

Comment:
I see `ThreadBlockInVM` makes the current thread stack walkable. But how can our previous implementation work without this guard?

Reply:
Previously, `MutexLocker` does "safepoint check".  That is, if safepoint synchronization is in progress, it will yield when attempting to acquire the lock (i.e. there is a `ThhreadBlockInVM` hidden in `Monitor::lock` in the code path of yielding).  That worked, err... to a certain degree.

It looks like if the thread that tries to acquire the lock yields, other threads cannot acquire the same lock, either.  That prevented `release_mutator` from acquiring the `gc_lock`.  That was the cause of the first source of deadlock described [here](https://github.com/mmtk/mmtk-openjdk/issues/278#issuecomment-2129928182)  We didn't have this problem before we removed the coordinator thread.  I think that was because when a resumed mutator runs so fast that it triggers GC, it acquires a different lock, and that won't try to acquire the `WorkerMonitor` currently held by `release_mutator`.  When another VM operation, say RevokeBias, triggers another STW, the Java threads simply yield, and the VM threads do the VM operation and release the Java threads.  No deadlock.  The deadlock only happens if the mutator is blocked on the `WorkerMonitor` in `GCRequester::request` which is *not* executed in saferegion.